Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -260,10 +260,28 @@ FileSystemModel::FileSystemModel(String root_path, Mode mode)
         dbgln("Event at \"{}\" on Node {}: {}", node.full_path(), &node, event);
 
         // FIXME: Your time is coming, un-granular updates.
-        node.has_traversed = false;
-        node.mode = 0;
-        node.children.clear();
-        node.reify_if_needed();
+        auto refresh_node = [](Node& node) {
+            node.has_traversed = false;
+            node.mode = 0;
+            node.children.clear();
+            node.reify_if_needed();
+        };
+
+        if (event.type == Core::FileWatcherEvent::Type::Deleted) {
+            auto canonical_event_path = LexicalPath::canonicalized_path(event.event_path);
+            if (m_root_path.starts_with(canonical_event_path)) {
+                // Deleted directory contains our root, so navigate to our nearest parent.
+                auto new_path = LexicalPath(m_root_path).dirname();
+                while (!Core::File::is_directory(new_path))
+                    new_path = LexicalPath(new_path).dirname();
+
+                set_root_path(new_path);
```

Comment:
Couldn't you just go to the parent directory of `canonical_event_path`, since `canonical_event_path` and all of it's children have been deleted? This would remove the need for the `while` loop.

Reply:
@MaxWipfli Possibly. I think my thought was, the parent might also have been deleted too (or that parent's parent, etc) - I'm not familiar with how file watchers work and which directories they notify you about.

Reply:
Having now tested this, just using the event path's parent does not work. If it's we're in `a/b/c`, and `b` is deleted, then it navigates to `b` and then fails and navigates to `a` but with a big error message where the directory contents should be displayed.

But testing this out did remind me there was one more thing I wanted to change, so thanks!